Added sorting functionality

TheGooner93 opened the pull request:
This is enabled via clicks on the column header, for all the current columns

vikaskyadav commented:
Hi, Please attach the screenshot also. Also, there might be some change of logic as there won't be any values available in firebase data. I can give some thinking lines regarding this.

TheGooner93 commented:
![image](https://user-images.githubusercontent.com/8807030/49026638-5b278200-f1c4-11e8-8fce-f2b2b98c9d07.png)

Changes:
1. Each column header is now sort friendly, on click.
2. Wordcloud below that filters the table based on clicked events

TheGooner93 commented:
Hi @vikaskyadav ,
Thanks for the suggestions, here are my followup queries:
1. For the first point suggested by you: if this is to be implemented as another column, would you want it to look like a "cloud" or just like a list of words(like in a column of the table) ?

2. For your the second point, currently I am assuming that everytime there is a new event added, OR when there is an event item clicked, I am updating a "weight" or "popularityIndex" for each individual event item in the firestore and based on its value iI'm associating a font size to the word cloud item. Question from my side is, do you want the exact same functionality done here, for each word of the event, and not for the whole event name itself?

vikaskyadav commented:
Word cloud doesn't need to look like a cloud. It is just a phrase to depict a collection of words. The column having wordcloud can depict `keywords` in anyway you want. Just it should not look bad. 

And, we can't go as per point 2 as you mentioned because `keywords` are the points of emphasis.
For ex: A Javascript Conference may have `keywords` as ["javascript","React","Angular","Webcomponents"]. We need to keep count of each keyword.

For your ease, you can use lodash, as I mentioned earlier.

Also, when a user will be filling form for a new event, one more field will be present there which will have all these keywords, plus the user should be able to add new keywords also.

Let me know if you have any difficulty understanding this.

vikaskyadav commented:
@TheGooner93  any updates on this?

TheGooner93 commented:
@vikaskyadav , sorry I've been caught up with work.. Can I give you an update in the weekend , maybe Sunday?

TheGooner93 commented:
Hi @vikaskyadav have a look at this commit, I've updated the look and feel of the wordcloud as a row of trending words, and fixed the Redux devtools related bug.

![image](https://user-images.githubusercontent.com/8807030/49288320-7517dc00-f4c6-11e8-84f5-3b58295cd8ed.png)

vikaskyadav commented:
@TheGooner93 The layout is good but as we discussed earlier, we need to remove the logic from event name.  Here is how the schema will look like. The `tags` field will have all the keywords in all events. We need to do counting on all such different keywords. Your logic is correct, just need a little updation.

`{
  "id": "UH5dDydLyDxdW7nr1M8Y",
  "dateFrom": {
    "seconds": 1542153600,
    "nanoseconds": 0
  },
  "dateTo": {
    "seconds": 1542240000,
    "nanoseconds": 0
  },
  "description": "This is for all the cloud computing enthhusiasts around the world.",
  "keyword": "",
  "name": "AWS Reinvent",
  "popularity": null,
  "tags": [
    {
      "key": 0,
      "label": "AWS"
    },
    {
      "key": 1,
      "label": "CLOUD"
    },
    {
      "key": 2,
      "label": "CLOUD COMPUTING"
    },
    {
      "key": 3,
      "label": "NETWORKING"
    },
    {
      "key": 4,
      "label": "NETWORK SECURITY"
    }
  ],
  "url": "https://awsreinvent.com/",
  "venue": "Online"
}`

TheGooner93 commented:
ok so, the `tags` field will be populated by us, based on user input? Like do we pull out the words ourselves whenever a new event is added? Whats the approach?

vikaskyadav commented:
yeah. whoever will submit the event will fill in the relevant tags also. We don't need to make any additional APi call for the tags. Whenever wee are loading the table data,we will get the tags also and we'll extract all the `tags` from the events, keeping a count of them and then we will render it in the wordcloud.

The count will become the popularity index and the text will be the label inside the `tag`

TheGooner93 commented:
I just had a look at the implementation by @geekyrajshri ; looks good. 
My only doubt, before implementing the changes to the wordcloud, is can we take the popularityIndex that is maintained on each event or should the popularity be associated with each tag?

My idea is to let there be a popularityIndex that is defined on the event level, be used to consider the priority of the corresponding tags that are part of the event record in firebase, and use this same popularityIndex to sort them in the wordcloud.

What do you think, @vikaskyadav ?

vikaskyadav commented:
Yeah, but that might not be a good idea to set popularityIndex corresponding to every tag because of some of the reasons:

-  There is no limit on number of tags and in reality there people can fill in anything.

- It might be the case that we may have to remove any event after it has been posted.

- It might also be the case that we bring in a functionality for user to change/remove tags for events that he/she  has added.

- Implementing your logic will incur more api calls for priority calculation which I would like to avoid.

**Suggestions:*

- We can make compute `popularityIndex` every time whenever we are rendering the table from fetched firebase data and compute it.

- This will keep popularityIndex almost in **Real-Time** and also it will be efficient also as number of API calls will still be same.

- Computing

TheGooner93 commented:
Yes, I wont set the popularityIndex onto the TAGS, but onto the EVENT that contains the tags. This way, each tags will be associated with the same popularityIndex as that of the parent event.  If you check the following image, you can understand that the popularityIndex is associated to the event.
![image](https://user-images.githubusercontent.com/8807030/49488821-53c04280-f86e-11e8-9abf-009cc7d3108d.png)

The popularityIndex wont have any additional API calls, as it created when an event is added for the first time, and updated when an event is clicked in the table. 

So ,to summarise my approach: at the time of adding events, or clicking on events in the events table, the popularityIndex of the **event** will be updated. The Wordcloud will show only tags that have been added for the event(present as part of event schema), sorted on the basis of popularityIndex of the **event**

Let me know what you think.

TheGooner93 commented:
By one time thing I mean, that the popularity of the event will only be added/changed at the time of event creation

vikaskyadav commented:
> Okay. So shall we avoid the clicking of events to affect the popularityIndex?

Yes, It shouldn't affect.

>By one time thing I mean, that the popularity of the event will only be added/changed at the time of event creation

I don't want to keep popularity as static count. It should be a dynamic count that might change anytime.

TheGooner93 commented:
How do you propose it should be made dynamic? Because right now, as per our discussion, the only time that it can be affected, is when the event is added by the user..

vikaskyadav commented:
the count can be computed on the client side whenever the data is fetched. Assume that there is nothing like count stored.

Ex;

There are 5 events coming in from firebase each having 2-3 tags:

event1: html, css
event2: javascript, php
event3: ruby, ruby on rails, backend,javascript
event4: html, css, javascript
event5: reactjs

then counts will be as : 
html:2
css:2
javascript:3
reactjs:1
php:1
ruby:1
ruby on rails:1
backend:1

and then priorities should be set accordingly. And This might change anytime.

TheGooner93 commented:
Oh I get your point now, that is why you were proposing a count and not a popularityIndex, guess we were both confusing each other!

TheGooner93 commented:
Hi @vikaskyadav made the changes as requested, have a look: now I'm looking at the counts of the tags and sorting the wordcloud based on it; I've removed the whole popularityIndex implementation from before.

TheGooner93 commented:
Hey, any updates on this?

TheGooner93 commented on src/EventWordCloud.js:
```diff
@@ -1,49 +1,93 @@
+          events[i].tags &&
+          events[i].tags.length &&
+          events[i].tags.map(tag => tag.label);
+        for (var j = 0; j < eventWords.length; j++) {
```
I dont think it will enter the loop itself, if eventWords is empty

vikaskyadav replied:
just try and check once. It isn't working.  If there isn't any keyword coming from firebase. it shouldn't.i.e. no keyword is present. It shouldn't show empty word cloud. And, also I'll send kudos for this PR as you have invested so much effort in this.

TheGooner93 replied:
ok will keep you posted

vikaskyadav commented:
Awesome work @TheGooner93 . :+1:

TheGooner93 commented:
@vikasyadav, yes I would like to work on it.

vikaskyadav commented:
Great! Let me prioritize the issues and raise them on Github first.
